### DISCUSSION TITLE
How to return a python list of number without modify the cpp class definitions?

### DISCUSSION THREAD

**Discussion by foreverlms** [pybind/pybind11#3638] [General]
Suppose I have a class:
```
class Foo{
public：
void* data;
}
```
So the data pointer will be filled with numbers (float/int...)，I export this class(m is the module I defined):
```
py::class_<Foo> foo(m,"Foo");
foo.def()// blabla
```

How can I define a member method in python `Foo` to return the data in pointer `data` without modify c++ `Foo`? thanks.
Just like:
```
foo = Foo()
result = foo.get_data() # type : list
```

**Comment by jiwaszki** [comment]
If you do not want to modify C++ you always can create a wrapper around the original class:

```cpp
class FooWrapper
{
public:
    FooWrapper(void *data, py::dtype type, std::vector<size_t> shape) : _type(type), _shape(shape)
    {
        Foo tmp_foo;
        _foo = tmp_foo;
        _foo.data = data;
    }

    Foo _foo;
    py::dtype _type;
    std::vector<size_t> _shape;
};
```

And create class bindings similar to this:

```cpp
py::class_<FooWrapper, std::shared_ptr<FooWrapper>> cls(module, "Foo");

cls.def(py::init([](py::array &data)
                    {
                        std::vector<size_t> shape(data.shape(), data.shape() + data.ndim());
                        return FooWrapper(const_cast<void *>(data.data()), data.dtype(), shape);
                    }),
        py::arg("data"));

cls.def_property_readonly(
    "data", [](FooWrapper &self)
    { return py::array(self._type,
                        self._shape,
                        self._foo.data,
                        py::cast(self)); });
```

If in your case you have no other means of extracting info like size/number of elements/byte size of element about void* from your class... I think it would be non-trivial or nearly impossible to get information about array you want. You need to have at least number of elements you are holding under void*. If there is any way to obtain them, you should return "raw data" -- that means flat list with shape of number of elements and type of uint8 for example.

**Reply by foreverlms** [reply]
@jiwaszki You are right. I forget the shape info about foo class. In fact, I want to connect Python and C++ with some class like `Tensor`. Thanks for your reply, I will have a try and reply to you as soon as possible. It semms good enough to solve my problem.
